feature/experimental/p4rt/ate_tests/traceroute_packetin_test/traceroute_packetin_test.go

```diff
@@ -0,0 +1,377 @@
+	top.Push(t).StartProtocols(t)
+
+	leader := p4rt_client.P4RTClient{}
+	if err := leader.P4rtClientSet(dut.RawAPIs().P4RT(t)); err != nil {
+		t.Fatalf("Could not initialize p4rt client: %v", err)
+	}
+
+	follower := p4rt_client.P4RTClient{}
+	if err := follower.P4rtClientSet(dut.RawAPIs().P4RT(t)); err != nil {
+		t.Fatalf("Could not initialize p4rt client: %v", err)
+	}
+
+	time.Sleep(120 * time.Second)
+	args := &testArgs{
+		ctx:      ctx,
+		leader:   &leader,
+		follower: &follower,
+		dut:      dut,
+		ate:      ate,
+		top:      top,
+	}
+
+	if err := setupP4RTClient(ctx, args); err != nil {
+		t.Fatalf("Could not setup p4rt client: %v", err)
+	}
+
+	args.packetIO = getTracerouteParameter(t)
+	testPacketIn(ctx, t, args, true)	// testPacketin for Ipv4 
+	testPacketIn(ctx, t, args, false)	// testPacketin for Ipv6
+}
+
+type TraceroutePacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+// GetTableEntry creates wbb acl entry related to Traceroute protocol.
+func (traceroute *TraceroutePacketIO) GetTableEntry(delete bool, IsIpv4 bool) []*wbb.ACLWbbIngressTableEntryInfo {
+	if IsIpv4 {
+		actionType := p4_v1.Update_INSERT
+		if delete {
+			actionType = p4_v1.Update_DELETE
+		}
+		return []*wbb.ACLWbbIngressTableEntryInfo{{
+			Type:    actionType,
+					IsIpv4:  0x1,
+					TTL:     0x1,
+					TTLMask: 0xFF,
+		},
+		{
+			Type:    actionType,
+					IsIpv4:  0x1,
+					TTL:     0x0,
+					TTLMask: 0xFF,
+		},
+	}
+	}	else {
+			actionType := p4_v1.Update_INSERT
+		if delete {
+			actionType = p4_v1.Update_DELETE
+		}
+		return []*wbb.ACLWbbIngressTableEntryInfo{{
+			Type:    actionType,
+					IsIpv6:  0x1,
+					TTL:     0x1,
+					TTLMask: 0xFF,
+		},
+		{
+			Type:    actionType,
+					IsIpv6:  0x1,
+					TTL:     0x0,
+					TTLMask: 0xFF,
+		},}
+		}
+}
+
+// GetPacketTemplate returns expected packets in PacketIn.
+func (traceroute *TraceroutePacketIO) GetPacketTemplate() *PacketIOPacket {
+	return &traceroute.PacketIOPacket
+}
+
+func (traceroute *TraceroutePacketIO) GetTrafficFlow(ate *ondatra.ATEDevice, isIpv4 bool,TTL uint8,  frameSize uint32, frameRate uint64) []*ondatra.Flow {
+	ethHeader := ondatra.NewEthernetHeader()
+	ipv4Header := ondatra.NewIPv4Header().WithSrcAddress(atePort1.IPv4).WithDstAddress(dutPort1.IPv4).WithTTL(uint8(TTL)) //ttl=1 is traceroute/lldp traffic
+	ipv6Header := ondatra.NewIPv6Header().WithSrcAddress(atePort1.IPv6).WithDstAddress(dutPort1.IPv6).WithHopLimit(uint8(TTL))
+	if isIpv4 {
+		flow := ate.Traffic().NewFlow("IP4").WithFrameSize(frameSize).WithFrameRateFPS(frameRate).WithHeaders(ethHeader, ipv4Header)
+		return []*ondatra.Flow{flow} 
+	}	else {
+		flow := ate.Traffic().NewFlow("IP6").WithFrameSize(frameSize).WithFrameRateFPS(frameRate).WithHeaders(ethHeader, ipv6Header)
+		return []*ondatra.Flow{flow}
+	}
+}
+
+
+
+// GetEgressPort returns expected egress port info in PacketIn.
+func (traceroute *TraceroutePacketIO) GetEgressPort() []string {
+	return []string{"0"}
```

Comment:
Based on p4info file, this is the target_egress_port. In your testcase, it should be the port-id related to dutPort1.

Reply:
@ArpitJoshi18  could you please make required changes on this as well?

Reply:
@mingyangcisco  Incorporated in 10/10 commit. Thanks for your input.